File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/noughtsandcrosses/data/NoughtsAndCrossesDocument.java
package com.zwstudio.logicpuzzlesandroid.puzzles.noughtsandcrosses.data;

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocument;
import com.zwstudio.logicpuzzlesandroid.common.data.MoveProgress;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.puzzles.noughtsandcrosses.domain.NoughtsAndCrossesGame;
import com.zwstudio.logicpuzzlesandroid.puzzles.noughtsandcrosses.domain.NoughtsAndCrossesGameMove;

import org.androidannotations.annotations.EBean;

/**
 * Created by zwvista on 2016/09/29.
 */

@EBean
public class NoughtsAndCrossesDocument extends GameDocument<NoughtsAndCrossesGame, NoughtsAndCrossesGameMove> {
    protected void saveMove(NoughtsAndCrossesGameMove move, MoveProgress rec) {
        rec.row = move.p.row;
        rec.col = move.p.col;
        rec.strValue1 = String.valueOf(move.obj);
    }
    public NoughtsAndCrossesGameMove loadMove(MoveProgress rec) {
        return new NoughtsAndCrossesGameMove() {{
            p = new Position(rec.row, rec.col);
            obj = rec.strValue1.charAt(0);
        }};
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/noughtsandcrosses/data/NoughtsAndCrossesDocument.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.noughtsandcrosses.data

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocument
import com.zwstudio.logicpuzzlesandroid.common.data.MoveProgress
import com.zwstudio.logicpuzzlesandroid.common.domain.Position
import com.zwstudio.logicpuzzlesandroid.puzzles.noughtsandcrosses.domain.NoughtsAndCrossesGame
import com.zwstudio.logicpuzzlesandroid.puzzles.noughtsandcrosses.domain.NoughtsAndCrossesGameMove
import org.androidannotations.annotations.EBean

@EBean
class NoughtsAndCrossesDocument : GameDocument<NoughtsAndCrossesGame, NoughtsAndCrossesGameMove>() {
    override fun saveMove(move: NoughtsAndCrossesGameMove, rec: MoveProgress) {
        rec.row = move.p.row
        rec.col = move.p.col
        rec.strValue1 = move.obj.toString()
    }

    override fun loadMove(rec: MoveProgress) =
        NoughtsAndCrossesGameMove(Position(rec.row, rec.col), rec.strValue1!![0])
}